src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h

```diff
@@ -0,0 +1,173 @@
+
+  	/** Setter for the training data X */
+  	virtual void set_training_data(
+  		Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic> * X_new,
+  		Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y_new
+  	);
+
+  	/** Setter for the trainable parameters of the cost function */
+  	virtual void set_trainable_parameters(
+  		Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* new_params
+  	);
+
+  	/** Setter for the cost function definition using stan */
+  	virtual void set_ith_cost_function(std::vector< std::function < stan::math::var(int32_t) > >* new_cost_f);
+
+    /** Setter for the overall cost function */
+    virtual void set_cost_function(std::function < stan::math::var(std::vector<stan::math::var>*) >* total_cost);
+
+
+  	~FirstOrderSAGCostFunctionInterface();
+
+  	/** Initialize to generate a sample sequence
+  	 *
+  	 */
+  	virtual void begin_sample();
+
+  	/** Get next sample
+  	 *
+  	 * @return false if reach the end of the sample sequence
+  	 * */
+  	virtual bool next_sample();
+
+  	/** Get the SAMPLE gradient value wrt target variables
+  	 *
+  	 * WARNING
+  	 * This method does return
+  	 * \f$ \frac{\partial f_i(w) }{\partial w} \f$,
+  	 * instead of
+  	 * \f$\sum_i{ \frac{\partial f_i(w) }{\partial w} }\f$
+  	 *
+  	 * For least squares cost function, that is the value of
+  	 * \f$\frac{\partial f_i(w) }{\partial w}\f$ given \f$w\f$ is known
+  	 * where the index \f$i\f$ is obtained by next_sample()
+  	 *
+  	 * @return sample gradient of variables
+  	 */
+  	virtual SGVector<float64_t> get_gradient();
+
+  	/** Get the cost given current target variables
+  	 *
+  	 * For least squares, that is the value of \f$f(w)\f$.
+  	 *
+  	 * @return cost
+  	 */
+  	virtual float64_t get_cost();
+
+    /** Get the sample size
+     *
+     * @return the sample size
+     */
+    virtual int32_t get_sample_size();
+
+    /** Get the AVERAGE gradient value wrt target variables
+     *
+     * Note that the average gradient is the mean of sample gradient from get_gradient()
+     * if samples are generated (uniformly) at random.
+     *
+     * WARNING
+     * This method returns
+     * \f$ \frac{\sum_i^n{ \frac{\partial f_i(w) }{\partial w} }}{n}\f$
+     *
+     * For least squares, that is the value of
+     * \f$ \frac{\frac{\partial f(w) }{\partial w}}{n} \f$ given \f$w\f$ is known
+     * where \f$f(w)=\frac{ \sum_i^n{ (y_i-w^t x_i)^2 } }{2}\f$
+     *
+     * @return average gradient of target variables
+     */
+    virtual SGVector<float64_t> get_average_gradient();
+
+
+  protected:
+  	/** X is the training data in column major matrix format */
+  	Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* m_X;
+
+  	/** y is the ground truth, or the correct prediction */
+  	Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* m_y;
+
+  	/** trainable_parameters are the variables that are optimized for */
+  	Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* m_trainable_parameters;
```

Comment:
this currently is foobar... you wont be able to serialize reproducible this class